src/python/pants/engine/rule_visitor.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import ast
+import inspect
+import itertools
+import logging
+import sys
+from functools import partial
+from typing import List
+
+from pants.engine.internals.selectors import AwaitableConstraints, GetParseError
+from pants.util.memo import memoized
+
+logger = logging.getLogger(__name__)
+
+
+def _is_awaitable_constraint(call_node) -> bool:
+    return isinstance(call_node.func, ast.Name) and call_node.func.id in ("Get", "Effect")
+
+
+def _get_starting_indent(source):
+    """Used to remove leading indentation from `source` so ast.parse() doesn't raise an
+    exception."""
+    if source.startswith(" "):
+        return sum(1 for _ in itertools.takewhile(lambda c: c in {" ", b" "}, source))
+    return 0
+
+
+def _get_lookup_names(attr: ast.Attribute | ast.Name):
+    names = []
+    while isinstance(attr, ast.Attribute):
+        names.append(attr.attr)
+        attr = attr.value
+    # NB: attr could be a constant, like `",".join()`
+    names.append(getattr(attr, "id", None))
+    return names
+
+
+class _AwaitableCollector(ast.NodeVisitor):
+    def __init__(self, func):
```

Comment:
Type hints are pretty annoying when dealing with the AST module. But it might be worth considering adding still to this function and the others